src/uct/ib/base/ib_md.c

```diff
@@ -838,13 +871,34 @@ static ucs_status_t uct_ib_mem_alloc(uct_md_h uct_md, size_t *length_p,
         goto err;
     }
 
-    length     = *length_p;
+    length = *length_p;
+
+#if HAVE_IBV_EXP_DM
+    if (flags & UCT_MD_MEM_FLAG_ON_DEVICE) {
+        memh->dm = uct_ib_mem_alloc_dm(md, length);
```

Comment:
seems that it would be simpler if uct_ib_md_reg_mr() call would be also inside uct_ib_mem_alloc_dm() - so uct_ib_mem_alloc_dm() would both allocate and register DM memory. uct_ib_md_reg_mr() on line 892 could be left unchanged then.

Reply:
not sure I got your idea... but implementation is updated